modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPackage.java

```diff
@@ -759,6 +760,30 @@ SimpleFeatureType correctSchema(SimpleFeatureType schema) {
         return builder.buildFeatureType();
     }
 
+    public void setupZM(FeatureEntry entry, SimpleFeatureType schema) {
+        GeometryDescriptor geomDescriptor = schema.getGeometryDescriptor();
+        if (geomDescriptor == null
+                || geomDescriptor.getUserData() == null
+                || !geomDescriptor.getUserData().containsKey(Hints.COORDINATE_DIMENSION)) {
+            return; // nothing to do
+        }
+        Integer dims = (Integer) geomDescriptor.getUserData().get(Hints.COORDINATE_DIMENSION);
+        if (dims == null || dims < 0 || dims > 4) {
+            return; // nothing to do
+        }
+        if (dims == 2) {
+            return; // default is 2 (xy)
+        }
+        if (dims == 3) {
+            entry.setZ(true); // assume 3rd is Z
+            return;
+        }
```

Comment:
After https://github.com/geotools/geotools/pull/5337 adds Hints.COORDINATE_MEASURES this method can avoid this assumption:
- dimensions=3, measures=0 = XYZ
- dimension=3, measure 1 = XYM